Revamp the Hours view

hawkrives opened the pull request:
Oh man. This PR has been a _long_ time in the making. (Sept. 13th is when I finally wrote down the [giant hours gist][1])

[1]: https://gist.github.com/hawkrives/9ffcf27347c08520542c33cd3a42fdd2#gistcomment-1897044

This is now a complete revamp of the styles and behavior of the Building Hours screen. 

The list itself no longer shows images, to increase information density and highlight the building status.

| before |     |
|:---:|:---:|
| <img width="480" alt="v1" src="https://cloud.githubusercontent.com/assets/464441/20739050/458cb87c-b67d-11e6-8424-218d862fbca8.png" /> | <img width="480" alt="v2" src="https://cloud.githubusercontent.com/assets/464441/20739051/458df430-b67d-11e6-96a5-25fef7c95730.png" />

| after |     |
|:---:|:---:|
| <img width="480" alt="android" src="https://cloud.githubusercontent.com/assets/464441/20739049/458bf2f2-b67d-11e6-863d-da76aefd927f.png" /> |  <img width="480" alt="ios" src="https://cloud.githubusercontent.com/assets/464441/20739048/4587977a-b67d-11e6-808b-892d98b62271.png" />

When tapped, a row will push a new BuildingDetail view onto the stack, exposing information such as the full title, gorgeous photograph, and complete schedules.

| before |     |
|:---:|:---:|
| <img width="480" alt="v1" src="https://cloud.githubusercontent.com/assets/464441/20739059/577666d2-b67d-11e6-8c22-42991e562a60.png" /> | <img width="480" alt="v2" src="https://cloud.githubusercontent.com/assets/464441/20739062/577f643a-b67d-11e6-9417-e117eeff4897.png" />

| after |     |
|:---:|:---:|
| <img width="480" alt="android" src="https://cloud.githubusercontent.com/assets/464441/20739061/577b8446-b67d-11e6-8448-e03f81e6b8fa.png" /> | <img width="480" alt="ios" src="https://cloud.githubusercontent.com/assets/464441/20739060/5779d4c0-b67d-11e6-92e1-8a6e7affef86.png" />


Speaking of schedules, this brings along an entirely new approach to the building schedules. Features of the new format include:

- multiple schedules per building
- multiple schedules per day
- buildings grouped by category
- notes displayed on schedules

A Building can either have a `schedule` key, containing a list of SingleBuildingScheduleType, or a `namedSchedules` key, containing a list of NamedBuildingScheduleType (which themselves contain a set of SingleBuildingScheduleType). 

A SingleBuildingScheduleType has a `days` key, containing a list of days of the week that it applies to, and a `from` and `to` key, in the format "h:mma" (12-hour:minute am/pm). It looks something like this:

```js
{days: [Mo, Tu, Fr], from: '1:00pm', to: '5:00pm'}
```

A NamedBuildingScheduleType has `title`, `hours`, and optional `notes` keys. `title` is the title of the schedule; `hours` is the list of SingleBuildingScheduleType; and `notes` are any notes about that schedule.

# Examples

- The Cage
- Stav Hall
- DiSCO
- Pool
- STORP


## The Cage
- Two schedules: one for the hot kitchen, one for else

| android | ios |
|:---:|:---:|
| <img width="480" alt="detail day and" src="https://cloud.githubusercontent.com/assets/464441/20739066/6749c874-b67d-11e6-9338-f0f4ec0d7576.png" /> | <img width="480" alt="detail day ios" src="https://cloud.githubusercontent.com/assets/464441/20739067/6755f342-b67d-11e6-9cb8-aab7b224d214.png" />

| android | ios |
|:---:|:---:|
| <img width="480" alt="detail late and" src="https://cloud.githubusercontent.com/assets/464441/20739068/675b6fc0-b67d-11e6-8963-243309a1c087.png" /> | <img width="480" alt="detail late ios" src="https://cloud.githubusercontent.com/assets/464441/20739072/6771f510-b67d-11e6-9bce-0f8f51dd592c.png" />

| android | ios |
|:---:|:---:|
| <img width="480" alt="list day and" src="https://cloud.githubusercontent.com/assets/464441/20739069/675e5c30-b67d-11e6-8340-50b2efceb0f1.png" /> | <img width="480" alt="list day ios" src="https://cloud.githubusercontent.com/assets/464441/20739070/6760bb06-b67d-11e6-8c95-a40e6604e3c6.png" />

| android | ios |
|:---:|:---:|
| <img width="480" alt="list late and" src="https://cloud.githubusercontent.com/assets/464441/20739071/67653924-b67d-11e6-861f-d4077ebd7ab1.png" /> | <img width="480" alt="list late ios" src="https://cloud.githubusercontent.com/assets/464441/20739073/67824eec-b67d-11e6-8f06-ba8efef509be.png" />


## Stav Hall
- Three schedules, breakfast, lunch, and dinner
- Note they are grouped separately when different days

| android | ios |
|:---:|:---:|
| <img width="480" alt="detail android" src="https://cloud.githubusercontent.com/assets/464441/20739082/6ddb2782-b67d-11e6-8a49-5c326ac9b28a.png" /> | <img width="480" alt="detail ios" src="https://cloud.githubusercontent.com/assets/464441/20739079/6dcedde2-b67d-11e6-9c48-e0463d21b5d3.png" />

| android | ios |
|:---:|:---:|
| <img width="480" alt="list android" src="https://cloud.githubusercontent.com/assets/464441/20739080/6dd32ac8-b67d-11e6-941d-ba087e8a2303.png" /> |  <img width="480" alt="list ios" src="https://cloud.githubusercontent.com/assets/464441/20739081/6dd74eb4-b67d-11e6-9fbd-48d4e281782e.png" />



## DiSCO / Rølvaag Ref Desk
- Two schedules for M-Th: 10-5 and 7-9


| android | ios |
|:---:|:---:|
| <img width="480" alt="detail android" src="https://cloud.githubusercontent.com/assets/464441/20739085/72c2bdb4-b67d-11e6-8052-1a70f675e94c.png" /> | <img width="480" alt="detail ios" src="https://cloud.githubusercontent.com/assets/464441/20739086/72c6a94c-b67d-11e6-95c7-13763fc0ccfe.png" />

| android | ios |
|:---:|:---:|
| <img width="480" alt="list android" src="https://cloud.githubusercontent.com/assets/464441/20739087/72c6f4d8-b67d-11e6-84c8-f114f4609703.png" /> | <img width="480" alt="list ios" src="https://cloud.githubusercontent.com/assets/464441/20739088/72c8d3a2-b67d-11e6-861e-eec2b0def615.png" />



## Skoglund Pool
- Four(!) schedules, for the different swim types.

| android | ios |
|:---:|:---:|
| <img width="480" alt="detail android" src="https://cloud.githubusercontent.com/assets/464441/20739085/72c2bdb4-b67d-11e6-8052-1a70f675e94c.png" /> | <img width="480" alt="detail ios" src="https://cloud.githubusercontent.com/assets/464441/20739086/72c6a94c-b67d-11e6-95c7-13763fc0ccfe.png" />

| android | ios |
|:---:|:---:|
| <img width="480" alt="list android" src="https://cloud.githubusercontent.com/assets/464441/20739087/72c6f4d8-b67d-11e6-84c8-f114f4609703.png" /> | <img width="480" alt="list ios" src="https://cloud.githubusercontent.com/assets/464441/20739088/72c8d3a2-b67d-11e6-861e-eec2b0def615.png" />



## STORP
- One schedule, but only M/Th/F

| android | ios |
|:---:|:---:|
| <img width="480" alt="android" src="https://cloud.githubusercontent.com/assets/464441/20739102/81039754-b67d-11e6-8fb7-5495acf56f99.png" /> | <img width="480" alt="ios" src="https://cloud.githubusercontent.com/assets/464441/20739101/8102d9b8-b67d-11e6-8043-f70f7d8dcec0.png" />

---

Sorry for the length! We'll hold off on merging this until @elijahverdoorn & co have signed off, too.

hawkrives committed:
add new building hours format

hawkrives committed:
add "categories" to the building hours

hawkrives committed:
fix summarizeDays to handle edge cases better

hawkrives committed:
add tests for summarizeDays

hawkrives committed:
migrate parseHours tests

hawkrives committed:
add test for normalizeBuildingSchedule

hawkrives committed:
update moment test helper

hawkrives committed:
add test for isBuildingOpenAtMoment

hawkrives committed:
rename formatStatusOfBuildingAtMoment to formatBuildingTimes

hawkrives committed:
add tests for formatBuildingTimes

hawkrives committed:
add optional moment-format to dayMoment helper

hawkrives committed:
add tests for getStatusOfBuildingAtMoment

hawkrives committed:
add tests for isBuildingOpen

hawkrives committed:
add tests for getShortBuildingStatus

hawkrives committed:
remove old tests

hawkrives committed:
potential getDetailedBuildingHours test

hawkrives committed:
rebuild BuildingHours

- now looks native-ish on both iOS and Android
- glancable, more compact list screen
- supports multiple schedules per building
  - and multiple schedules per day
- features the building image heavily in the new Detail screen
  - includes a parallax effect on the image as you scroll
  - the detail screen shows you the details of the schedule for that
    building

hawkrives committed:
add helper functions for working with the new data

hawkrives committed:
add flow typedef for rn-parallax-view

hawkrives committed:
remove old building test

hawkrives committed:
update tests for getDetailsBuildingSnapshot

hawkrives committed:
fix post office time format

hawkrives committed:
add hours for STORP, Wellness Center, Health Center

hawkrives committed:
add underlayColor to BuildingHours rows

hawkrives committed:
add "christmasfest" break type

hawkrives committed:
add named "late night" schedule to the Cage

hawkrives committed:
remove incorrect summer hours for the Health Center

hawkrives committed:
move notes to a per-schedule location

hawkrives committed:
add Schedule Notes to iOS and Android details screens

hawkrives committed:
de-bold the active schedule on the iOS detail view

hawkrives committed:
fix up the list icons on iOS

hawkrives committed:
add debugging

drewvolz commented on views/building-hours/building-hours-helpers.js:
```diff
@@ -0,0 +1,193 @@
+
+  // if the number of days given is not the number of days in the span,
+  // join the list. (There's no point to converting them here.)
+  if (endIndex - startIndex !== sortedDays.length - 1) {
```
Nice short-circuiting logic. I threw my own tests at it and concluded that it works. Really nice simplified logic with this line right here. 👍

hawkrives replied:
When I wrote that, I was half-thinking about other things, and it then took be a solid 20 minutes of thinking about it to understand it enough to write the comment… I kept missing the fact that the indices are taken from the master list of days, rather than the given set.

drewvolz commented on data/new-building-hours.js:
```diff
@@ -0,0 +1,716 @@
+  },
+
+  {
+    name: 'STORP',
```
I want to propose adding an expanded name, or a comment/detail field that displays inside of the building hours detail view. The purpose would be to clarify the names of acronyms or otherwise ambiguous names.

```
name: 'STORP',
detail: 'St. Olaf Outdoor Recreation Program',
```

hawkrives replied:
Done as of e75435e280b2558dff81010f216bab8a67e22620.

hawkrives committed:
add "Print Center"

hawkrives committed:
add Boe image

hawkrives committed:
add image for print center

hawkrives committed:
highlight chapel time under the Boe schedule

hawkrives committed:
use the image for the print center

hawkrives committed:
[wip] add "closedForCommunityTime" key

hawkrives committed:
[wip] begin work with an interimSchedule key

hawkrives committed:
add wellness image

hawkrives committed:
use the wellness image

hawkrives committed:
add SARN

hawkrives committed:
add support for "isPhysicallyOpen"

if a schedule is marked as "physically open: false", it is never
reported as being "open". This prevents it from being reported as open,
from being bolded as the active schedule, and from being reported as
"opens soon".

hawkrives committed:
remove interimSchedule support

hawkrives committed:
remove Chapel Time from Boe listing

hawkrives committed:
remove ability to have non-named schedules

"Hours" remains a special schedule name; it will be hidden from the list
view.

hawkrives committed:
add buntrock summer hours

hawkrives committed:
move closedForChapelTime to the schedule, not the building

hawkrives committed:
update tests to match new building input format

hawkrives committed:
remove normalizeBuildingSchedule tests

hawkrives committed:
add chapelSchedule to building hours

hawkrives committed:
add support for closedForChapelTime

this returns 'Closed (Chapel)' or 'Closed for chapel' during chapel
and/or community time, as defined in the `chapelSchedule` set in the
building hours file.

hawkrives committed:
add basic test for isChapelTime

hawkrives committed:
extract the dayOfWeek typecast into a function

hawkrives committed:
add chapel times to the "Closed for chapel" message

hawkrives committed:
add subtitle and abbreviation support

hawkrives committed:
add 'chapel' to the list of building status types

hawkrives committed:
ask flow to restrict us from using any extraneous keys

hawkrives committed:
add ASC hours

hawkrives committed:
move the IT Helpdesk to the "Help and Support" category

hawkrives committed:
update DiSCO subtitle

@drewvolz gave me the full name

drewvolz reviewed (approved):
Really terrific work. Nothing stuck out, but I will leave final review to @zimmer18 and @elijahverdoorn. @ekmanzu  and @mattk410 should look at this, too.
